include/crocoddyl/core/control-base.hpp

```diff
@@ -0,0 +1,187 @@
+   */
+  virtual void calcDiff(const boost::shared_ptr<ControlParametrizationDataAbstract>& data, double t,
+                        const Eigen::Ref<const VectorXs>& p) const = 0;
+
+  // virtual MatrixXs calcDiff_J(double t, const Eigen::Ref<const VectorXs>& p) const;
+
+  /**
+   * @brief Compute the product between a specified matrix and the Jacobian of the control (with respect to the
+   * parameters)
+   *
+   * @param[in]  t      Time
+   * @param[in]  p      Control parameters
+   * @param[in]  A      A matrix to multiply times the Jacobian
+   * @param[out] out    Product between the matrix A and the Jacobian of the control with respect to the parameters
+   */
+  virtual void multiplyByJacobian(double t, const Eigen::Ref<const VectorXs>& p, const Eigen::Ref<const MatrixXs>& A,
+                                  Eigen::Ref<MatrixXs> out) const = 0;
+
+  virtual MatrixXs multiplyByJacobian_J(double t, const Eigen::Ref<const VectorXs>& p,
+                                        const Eigen::Ref<const MatrixXs>& A) const;
+
+  /**
+   * @brief Compute the product between the transposed Jacobian of the control (with respect to the parameters) and
+   * a specified matrix
+   *
+   * @param[in]  t      Time
+   * @param[in]  p      Control parameters
+   * @param[in]  A      A matrix to multiply times the Jacobian
+   * @param[out] out    Product between the transposed Jacobian of the control with respect to the parameters and the
+   * matrix A
+   */
+  virtual void multiplyJacobianTransposeBy(double t, const Eigen::Ref<const VectorXs>& p,
+                                           const Eigen::Ref<const MatrixXs>& A, Eigen::Ref<MatrixXs> out) const = 0;
+
+  virtual MatrixXs multiplyJacobianTransposeBy_J(double t, const Eigen::Ref<const VectorXs>& p,
+                                                 const Eigen::Ref<const MatrixXs>& A) const;
+  /**
+   * @brief Return the dimension of the control value
+   */
+  std::size_t get_nu() const;
+
+  /**
+   * @brief Return the dimension of control parameters
+   */
+  std::size_t get_np() const;
+
+ protected:
+  std::size_t nu_;  //!< Control dimension
+  std::size_t np_;  //!< Control parameters dimension
+};
+
+template <typename _Scalar>
+struct ControlParametrizationDataAbstractTpl {
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  typedef _Scalar Scalar;
+  typedef MathBaseTpl<Scalar> MathBase;
+  typedef typename MathBase::VectorXs VectorXs;
+  typedef typename MathBase::MatrixXs MatrixXs;
+
+  template <template <typename Scalar> class Model>
+  explicit ControlParametrizationDataAbstractTpl(Model<Scalar>* const model)
+      : u_diff(model->get_nu()), u_params(model->get_np()), J(model->get_nu(), model->get_np()) {
+    u_diff.setZero();
+    u_params.setZero();
+    J.setZero();
+  }
+  virtual ~ControlParametrizationDataAbstractTpl() {}
+
+  VectorXs u_diff;    //!< value of the control
+  VectorXs u_params;  //!< value of the control parameters
```

Comment:
Actually in that discussion, which involved also @proyan , we came to the conclusion of using the verbose names u_diff and u_params when referring to the vectors themselves, while using the shorter names w and u when referring to derivatives of quantities (such as cost or dynamics) with respect to u_diff and u_params.

Reply:
My understanding of the discussion is:
  1. @andreadelprete suggested to use a mix of short and long names in https://github.com/loco-3d/crocoddyl/pull/960#issuecomment-865679096
  2. @proyan didn't have any preference for `u_diff`, even he suggested what it could be considered a short name in https://github.com/loco-3d/crocoddyl/pull/960#issuecomment-866705880
  3. I proposed to use `w` as name in https://github.com/loco-3d/crocoddyl/pull/960#issuecomment-867039298, which it is indeed a short name as well. 

Unfortunately, mixing short and and long names as proposed in 1 is very confusing. Indeed, I struggled many times when I was reviewing the code. Examples of my struggles were: 
  - I wasn't clear what was the correct "control" to use in different parts of the code (but mainly in the integrators)
  - I confused many times if `w` (e.g., in `data->Lwu`) refer to the diff action or the control parametrization.

I would highly advice to do not follow the road proposed in 1. I really like much to have a single letter for everything. Otherwise, it starts to be a bit of *spaghetti code*. Reason why I would also highly advice to never break this rule of thumb.

Reply:
I agree with Andrea that using long descriptive names for vectors and shorter names for derivs would be easier. `u_diff` could explain more than `w` and demand less questions in the future.
But we are dragging out this point, I don't think it is that big an issue.

Reply:
After going through the code for a while I agree with Carlos in the end. Having the two names is a bit confusing, so I've used `w` and `u` everywhere in the new version of the code.